devito/types/petsc.py

```diff
@@ -0,0 +1,111 @@
+from devito.types.basic import AbstractFunction
+from devito.tools import dtype_to_petsctype, CustomDtype
+import numpy as np
+from devito.types import LocalObject
+
+
+class DM(LocalObject):
+    """
+    PETSc Data Management object (DM).
+    """
+    dtype = CustomDtype('DM')
+
+
+class Mat(LocalObject):
+    """
+    PETSc Matrix object (Mat).
+    """
+    dtype = CustomDtype('Mat')
+
+
+class Vec(LocalObject):
+    """
+    PETSc Vector object (Vec).
+    """
+    dtype = CustomDtype('Vec')
+
+
+class PetscMPIInt(LocalObject):
+    """
+    PETSc datatype used to represent ‘int’ parameters
+    to MPI functions.
+    """
+    dtype = CustomDtype('PetscMPIInt')
+
+
+class KSP(LocalObject):
+    """
+    PETSc KSP : Linear Systems Solvers.
+    Manages Krylov Methods.
+    """
+    dtype = CustomDtype('KSP')
+
+
+class PC(LocalObject):
+    """
+    PETSc object that manages all preconditioners (PC).
+    """
+    dtype = CustomDtype('PC')
+
+
+class KSPConvergedReason(LocalObject):
+    """
+    PETSc object - reason a Krylov method was determined
+    to have converged or diverged.
+    """
+    dtype = CustomDtype('KSPConvergedReason')
+
+
+class PETScFunction(AbstractFunction):
+    """
+    PETScFunctions.
+    """
+    _data_alignment = False
+
+    def __init_finalize__(self, *args, **kwargs):
+
+        super(PETScFunction, self).__init_finalize__(*args, **kwargs)
+
+        self._is_const = kwargs.get('is_const', False)
+
+    @classmethod
+    def __dtype_setup__(cls, **kwargs):
```

Comment:
`cls` is never used and can be dropped from the args

Reply:
it must stay

Reply:
why do u need to override this method?

Reply:
Otherwise the `dtype` for each `PETScFunction` is `None` and I use the `dtype` inside the new function `dtype_to_petsctype`. I think this works for `PETScFunctions` because the petsctype will actually depend on the `dtype` of the object (e.g `Function`) provided by the user.